Add support for extended block lengths

spevs opened the pull request:
This PR adds the following:

* `parse_ieee_block_header` now supports decoding blocks ≥ 1 GB using hexadecimal length digit (Tektronix/LeCroy)
* New function `parse_rs_block_header` supports decoding blocks ≥ 1 GB using parentheses-delimited length (Rohde & Schwarz)
* New `header_fmt = "rs"` option added to `read/query_binary_values()`. Internally this first calls the function `parse_ieee_or_rs_block_header`, which detects whether the block length begins with a parentheses and then calls the `ieee` or `rs` parser as appropriate. This implementation is used because R&S only uses this format when the block length exceeds 1 GB, so it's best to use automatic detection rather than need to have the user explicitly know whether it is using the parentheses format or not
* Exposes the `length_before_block` and `raise_on_late_block` options to the user interface of `read/query_binary_values()`, to allow user to set these values. Previously these were not accessible by the user, and always set to 25 and False. Some Tektronix scopes can return trace metadata in ASCII in front of the block data (before the block header) so a late block header >25 bytes can be expected, but previously it will always raise a warning :
```
UserWarning: The beginning of the block has been found at 156 which is an unexpectedly large value. The actual block may have been missing a beginning marker but the block contained one:
bytearray(b'1;8;BINARY;RI;MSB;"Ch1, DC coupling, 1.000V/div, 200.0us/div, 400000 points, Sample mode";1000;Y;"s";5.0000E-9;0.0000;124880;"V";40.0000E-3;1.0000;0.0000;1;#41000@AA?@?A?@@AB?A@C@A@')
```
In future it will be useful to add some way of retrieving this message in front of the block header (currently it is thrown away) but that is not included in this PR.

As a note, since the `header_fmt = rs` option in this implementation will interpret both IEEE (including the hex digit extension) and RS formatted blocks, it is somewhat "universal" for definite-length blocks, but doesn't detect indefinite length blocks if they happen to start with `"("`. The R&S format is only used for definite length blocks so I didn't think this would be a concern. But the user should be careful not to specify `header_fmt = rs` if an indefinite block is expected.

For the hex extension, I added a check to make sure the block length makes sense if the digit is 0xA to detect if an HP block is being decoded with the IEEE parser (the block length should be much more than 64K if there are 0xA = 10 digits, if it's not then it must have been an HP block, and an exception is raised with a descriptive message). But I don't have a way of testing this since I don't have any instruments that return blocks in HP format.

I tested that the R&S block decoding works, and the regular IEEE block decoding also still works.

- [x] Closes #911
- [ ] Executed ``ruff check . && ruff format -c . --check`` with no errors
- [x] The change is fully covered by automated unit tests
- [ ] Documented in docs/ as appropriate
- [x] Added an entry to the CHANGES file

spevs commented:
I've added support for writes in the R&S format using `header_fmt = 'rs'`.

MatthieuDartiailh commented:
Perfect ! Could you just add a changelog entry and I will merge (sorry for not mentioning it earlier).

MatthieuDartiailh commented:
Actually Codecov just reminded that for the pieces of code under util we could/should have some tests (under pyvisa/testsuite).

spevs commented:
Also just to put it in writing, I also realized a correction about my original note:

> As a note, since the header_fmt = rs option in this implementation will interpret both IEEE (including the hex digit extension) and RS formatted blocks, it is somewhat "universal" for definite-length blocks, but doesn't detect indefinite length blocks if they happen to start with "(". The R&S format is only used for definite length blocks so I didn't think this would be a concern. But the user should be careful not to specify header_fmt = rs if an indefinite block is expected.

Actually it will handle indefinite block lengths too since they always start with a zero, not just the data straight away. So they will always be detected as "not starting with parentheses" and branch into the normal IEEE decoder and be handled there. So the "rs" header format option (for reads) should be able to actually handle anything except HP formatted blocks.

MatthieuDartiailh commented:
There are still some uncovered error paths. If you cannot add the relevant tests, I will have a look at it next week.

spevs commented:
Thanks. I added a test case for 537, although in normal circumstances I realized this is not reachable, because this function wouldn't be called directly, it would be called by `parse_ieee_or_rs_block_header`, but that function does its own check of the same in order to determine which way to branch. So the check is always guaranteed to pass in the child function.

Line 806 check is unneeded, since R&S format doesn't have the concept of an indefinite block length. I've removed it.

Line 809 I'm not sure why it's missed, I think it should have been covered already under `test_binary_block_shorter_than_advertized`. Anyway I added another test case to try to force it to reach.

MatthieuDartiailh commented on pyvisa/testsuite/test_util.py:
```diff
@@ -383,62 +444,106 @@ def fblock(block, cont):
-            (util.from_ieee_block, util.from_hp_block),
+            ("ieee", "hp", "rs"),
+            (util.to_ieee_block, util.to_hp_block, util.to_rs_block),
+            (util.from_ieee_block, util.from_hp_block, util.from_ieee_or_rs_block),
```
You should add a case using directly from_rs_block to catch the error on line 804.

spevs replied:
Done.

spevs commented on pyvisa/testsuite/test_util.py:
```diff
@@ -448,9 +558,14 @@ def test_weird_binary_block_header_raise(self):
-            ("ieee", "hp"),
-            (util.to_ieee_block, util.to_hp_block),
-            (util.from_ieee_block, util.from_hp_block),
+            ("ieee", "hp", "rs"),
```
Whoops :)
